File: src/contests/cg/puzzle/go/max-area.go
// 🎮 CodinGame Puzzle - max-area
// https://www.codingame.com/training/easy/max-area

package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	var n int
	fmt.Fscan(reader, &n)
	a := make([]int64, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(reader, &a[i])
	}
	left, right := 0, n-1
	var best int64 = 0
	for left < right {
		h := a[left]
		if a[right] < h {
			h = a[right]
		}
		area := h * int64(right-left)
		if area > best {
			best = area
		}
		if a[left] < a[right] {
			left++
		} else {
			right--
		}
	}
	fmt.Println(best)
}

File: src/contests/cg/puzzle/pascal/max-area.pas
// 🎮 CodinGame Puzzle - max-area
// https://www.codingame.com/training/easy/max-area

{$mode objfpc}{$H+}
var
  n, i, left, right: integer;
  a: array of int64;
  best, h, area: int64;
begin
  readln(n);
  setlength(a, n);
  for i := 0 to n - 1 do read(a[i]);
  readln;
  left := 0;
  right := n - 1;
  best := 0;
  while left < right do
  begin
    if a[left] < a[right] then h := a[left] else h := a[right];
    area := h * (right - left);
    if area > best then best := area;
    if a[left] < a[right] then inc(left) else dec(right);
  end;
  writeln(best);
end.
